gql/dsl.py

```diff
@@ -1,179 +1,378 @@
+        :class:`sync session <gql.client.SyncClientSession>`
+
+    :raises TypeError: if an argument is not an instance of :class:`DSLField`
+    :raises AssertionError: if an argument is not a field of a root type
+    """
+
+    # Check that we receive only arguments of type DSLField
+    # And that they are a root type
+    for field in fields:
+        if not isinstance(field, DSLField):
+            raise TypeError(
+                f"fields must be instances of DSLField. Received type: {type(field)}"
+            )
+        assert field.type_name in ["Query", "Mutation", "Subscription"], (
+            "fields should be root types (Query, Mutation or Subscription)\n"
+            f"Received: {field.type_name}"
+        )
+
+    # Get the operation from the first field
+    # All the fields must have the same operation
+    operation = fields[0].type_name.lower()
+
+    return DocumentNode(
+        definitions=[
+            OperationDefinitionNode(
+                operation=OperationType(operation),
+                selection_set=SelectionSetNode(
+                    selections=FrozenList(DSLField.get_ast_fields(fields))
+                ),
+            )
+        ]
+    )
+
+
+class DSLSchema:
+    """The DSLSchema is the root of the DSL code.
+
+    Attributes of the DSLSchema class are generated automatically
+    with the `__getattr__` dunder method in order to generate
+    instances of :class:`DSLType`
+    """
+
+    def __init__(self, schema: GraphQLSchema):
+        """Initialize the DSLSchema with the given schema.
+
+        :param schema: a GraphQL Schema provided locally or fetched using
+                       an introspection query. Usually `client.schema`
+        :type schema: GraphQLSchema
+
+        :raises TypeError: if the argument is not an instance of :class:`GraphQLSchema`
+        """
+
+        if not isinstance(schema, GraphQLSchema):
+            raise TypeError(
+                f"DSLSchema needs a schema as parameter. Received: {type(schema)}"
+            )
+
+        self._schema: GraphQLSchema = schema
+
+    def __getattr__(self, name: str) -> "DSLType":
+
+        type_def: Optional[GraphQLNamedType] = self._schema.get_type(name)
+
+        if type_def is None:
+            raise AttributeError(f"Type '{name}' not found in the schema!")
+
+        assert isinstance(type_def, GraphQLObjectType) or isinstance(
+            type_def, GraphQLInterfaceType
+        )
+
         return DSLType(type_def)
 
-    def query(self, *args, **kwargs):
-        return self.execute(query(*args, **kwargs))
 
-    def mutate(self, *args, **kwargs):
-        return self.query(*args, operation="mutation", **kwargs)
+class DSLType:
+    """The DSLType represents a GraphQL type for the DSL code.
+
+    It can be a root type (Query, Mutation or Subscription).
+    Or it can be an interface type (Character in the StarWars schema).
+    Or it can be an object type (Human in the StarWars schema).
```

Comment:
The root types are also object types, so maybe rephrase like this:

It can be a root type (Query, Mutation or Subscription).
Or it can be any other object type (Human in the StarWars schema).
Or it can be an interface type (Character in the StarWars schema).

Reply:
Done